articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,130 +38,466 @@ Result result = session.run(queryString);
+CREATE (n:`Fancy Name`);
 ----
 
-Even though we passed `$studentName` as a parameter to the query, the parameter is being appended into the query string for execution by `apoc.cypher.doIt()`.
-The same vulnerability would exist if we were processing a LOAD CSV query like before, and appending into a query string for execution by APOC.
+To use a backtick within such a name, it must be escaped using another backtick `&#96;`.
+
+For more information on escape characters, see the Cypher Manual on https://neo4j.com/docs/cypher-manual/current/syntax/expressions/[Expressions]
+and https://neo4j.com/docs/cypher-manual/current/syntax/naming/[Naming rules and recommendations].
+
+=== When Sanitization is Necessary
+
+Node labels, relationship types and parameters may contain non-alphabetic characters, including numbers, symbols and whitespace characters, but must be escaped using backticks.
+For example: `node label with spaces`.
+This means that when dynamically building a query using string concatenation, sanitization needs to be done on the escaping of backticks.
+In Cypher, a backtick is escaped using another backtick `&#96;&#96;`.
+For other types, for example string literals, that are opened and closed with either single quotes `'` or double quotes `"`,
+the sanitization would be done by escaping the quote character with a backslash `\`.
+Note however that where string literals can be used, so can parameters, and it is recommended to always parameterize instead of sanitize the input to avoid Cypher Injection.
+
+Here is an example of a simple dynamic label injection attack:
+
+`query = "MATCH (s:School)-[:IN]->(c:&#96;" + cityName + "&#96;) RETURN s;`
+
+With this query we want to search for all schools that are in a certain city, unfortunately our city names are Node Labels,
+so it isn't possible to parameterize the input.
+
+A possible attack input would be:
+
+`Input = &#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; \\`
+
+The backtick escapes the label name context and the parentheses closes the node.
+The `UNION` here then ensures that a match is made, as if the first `MATCH` statement doesn't return anything, the next part of the query won't be run.
+The `WITH` reduces the result set down to one row and then the final part will delete everything in the database.
+
+This attack was not possible to avoid using parameterization.
+To avoid this attack sanitization must be used.
+
+[Note]
+====
+The best protection against Cypher Injection is to always parameterize user input.
+If possible, update your data model to avoid needing to query using dynamic labels.
+In this example, the refactoring would be to move the city name to a parameter.
+
+`MATCH (c: City { name: $cityName }) RETURN name`
+====
+
+
+It is possible to add validation to the user input as well, in this case, validating the city name is a real city name before passing it into the database, and rejecting it otherwise.
```

Comment:
```suggestion
It is possible to add validation to the user input as well. In the above example, a real city name should probably not contain special characters such as a backtick or a colon. Validating the city name before passing it into the database, and rejecting suspicious input could help you to avoid cypher injection.
```

Plot twist: I would probably validate a city is real with a database call 😏. could we add this to the validation section? It feels like going back and forth between sanitization and validation otherwise.

Reply:
Hahaha noooo how could you!?!? :P 

I feel like it fits here personally